routing/triert/table.go

```diff
@@ -0,0 +1,227 @@
+	keysNext := trie.Remove(keys, tkey.Key(kk))
+
+	// if trie is unchanged then we didn't remove the key
+	if keysNext == keys {
+		return false, nil
+	}
+
+	// make a copy of keyNodes without removed key
+	// could avoid this if we held key/nodeid tuple in the trie
+	keyNodes := rt.keyNodes.Load().(map[string]address.NodeID)
+	if _, exists := keyNodes[string(kk)]; exists {
+		keyNodesNext := make(map[string]address.NodeID, len(keyNodes))
+		for k, v := range keyNodes {
+			if k == string(kk) {
+				continue
+			}
+			keyNodesNext[k] = v
+		}
+		rt.keyNodes.Store(keyNodesNext)
+	}
+
+	rt.keys.Store(keysNext)
+	return true, nil
+}
+
+// NearestPeers returns the n closest peers to a given key
+func (rt *TrieRT) NearestPeers(ctx context.Context, kk key.KadKey, n int) ([]address.NodeID, error) {
+	if kk.Size() != rt.self.Size() {
+		return nil, ErrKeyWrongLength
+	}
+
+	keys := rt.keys.Load().(*trie.Trie)
+	closestKeys, err := closestAtDepth(ctx, tkey.Key(kk), keys, 0, n, make([]tkey.Key, 0, n))
+	if err != nil {
+		return nil, err
+	}
+	keyNodes := rt.keyNodes.Load().(map[string]address.NodeID)
+
+	nodes := make([]address.NodeID, 0, len(closestKeys))
+	for _, c := range closestKeys {
+		if id, ok := keyNodes[string(key.KadKey(c))]; ok {
+			nodes = append(nodes, id)
+		}
+	}
+
+	return nodes, nil
+}
+
+func closestAtDepth(ctx context.Context, k tkey.Key, t *trie.Trie, depth int, n int, found []tkey.Key) ([]tkey.Key, error) {
+	if ctx.Err() != nil {
+		return found, ctx.Err()
+	}
+
+	// If we've already found enough peers, abort.
+	if n == len(found) {
+		return found, nil
+	}
+
+	// Find the closest direction.
+	dir := k.BitAt(depth)
+	var chosenDir byte
+	if t.Branch[dir] != nil {
+		// There are peers in the "closer" direction.
+		chosenDir = dir
+	} else if t.Branch[1-dir] != nil {
+		// There are peers in the "less closer" direction.
+		chosenDir = 1 - dir
+	} else if t.Key != nil {
+		// We've found a leaf
+		return append(found, t.Key), nil
+	} else {
+		// We've found an empty node?
+		return found, nil
+	}
+
+	var err error
+	// Add peers from the closest direction first, then from the other direction.
+	found, err = closestAtDepth(ctx, k, t.Branch[chosenDir], depth+1, n, found)
+	if err != nil {
+		return found, err
+	}
+
+	return closestAtDepth(ctx, k, t.Branch[1-chosenDir], depth+1, n, found)
+}
+
+func (rt *TrieRT) Find(ctx context.Context, kk key.KadKey) (address.NodeID, error) {
+	if kk.Size() != rt.self.Size() {
+		return nil, ErrKeyWrongLength
+	}
+
+	keyNodes := rt.keyNodes.Load().(map[string]address.NodeID)
+
+	if id, ok := keyNodes[string(key.KadKey(kk))]; ok {
+		return id, nil
+	}
+
+	return nil, nil
+}
+
+// Size returns the number of peers contained in the table.
+func (rt *TrieRT) Size() int {
+	keys := rt.keys.Load().(*trie.Trie)
+	return keys.Size()
+}
+
+// CplSize returns the number of peers in the table that share the specified common prefix length with the table's key.
+func (rt *TrieRT) CplSize(cpl int) int {
```

Comment:
Judging from the comments, this should do the same: https://github.com/libp2p/go-libp2p-xor/blob/1a5ca6fd64c49fc92364b47a005b1734d53bf5f8/kademlia/bucket.go#L10 and looks a bit more concise.

However, judging from the code, it seems that both are doing different things :D 

> // BucketAtDepth returns the bucket in the routing table at a given depth.
> // A bucket at depth D holds contacts that share a prefix of exactly D bits with node.

vs

> // CplSize returns the number of peers in the table that share the specified common prefix length with the table's key.

Reply:
Rewritten this function but I agree it's similar. But since we're moving the xor trie into the repo we won't be using the old xor/kademlia package